jsapp/js/account/stripe.utils.ts

```diff
@@ -127,3 +115,103 @@ export const getSubscriptionsForProductId = (
   }
   return null;
 };
+
+/*
+ * Performs logical operations to determine what information to provide about
+ * the upcoming status of user's subscription.
+ */
+export const getSubscriptionChangeDetails = (
+  currentPlan: SubscriptionInfo | null,
+  products: Product[]
+) => {
+  if (!(currentPlan && products.length)) {
+    return null;
+  }
+  let nextProduct: BaseProduct | null = null;
+  let date = '';
+  let type: SubscriptionChangeType = SubscriptionChangeType.NO_CHANGE;
+  if (currentPlan.cancel_at) {
+    date = currentPlan.cancel_at;
+    type = SubscriptionChangeType.CANCELLATION;
+  } else if (
+    currentPlan.schedule &&
+    currentPlan.schedule.status === 'active' &&
+    currentPlan.schedule.phases?.length &&
+    currentPlan.schedule.phases.length > 1
+  ) {
+    let nextPhaseItem = currentPlan.schedule.phases[1].items[0];
+    for (const product of products) {
+      let price = product.prices.find(
+        (price) => price.id === nextPhaseItem.price
+      );
+      if (price) {
+        nextProduct = product;
+        date = convertUnixTimestampToUtc(
+          currentPlan.schedule.phases[0].end_date!
+        );
+        if (nextProduct.id === currentPlan.items[0].price.product.id) {
+          if (currentPlan.quantity !== nextPhaseItem.quantity) {
+            type = SubscriptionChangeType.QUANTITY_CHANGE;
+          } else {
+            type = SubscriptionChangeType.PRICE_CHANGE;
+          }
+        } else {
+          type = SubscriptionChangeType.PRODUCT_CHANGE;
+        }
+        break;
+      }
+    }
+  } else if (currentPlan && type === SubscriptionChangeType.NO_CHANGE) {
+    date = currentPlan.current_period_end;
+    type = SubscriptionChangeType.RENEWAL;
+  }
+  return {nextProduct, date, type};
+};
+
+/**
+ * Takes a Stripe quantity (representing a total number of submissions included with a plan)
+ * and returns the transformed quantity. The total price of the transaction can be
+ * found by (transformed quantity x price unit amount).
+ * @param baseQuantity - the `quantity` field of the subscription in Stripe (total submission limit)
+ * @param transform - the `transform_quantity` field of the price
+ */
+export const getAdjustedQuantityForPrice = (
+  baseQuantity: number,
+  transform: TransformQuantity | null
+) => {
+  let adjustedQuantity = baseQuantity;
+  if (transform?.divide_by) {
+    adjustedQuantity /= transform.divide_by;
+  }
+  if (transform?.round === 'up') {
+    adjustedQuantity = Math.ceil(adjustedQuantity);
+  }
+  if (transform?.round === 'down') {
+    adjustedQuantity = Math.floor(adjustedQuantity);
+  }
+  return adjustedQuantity;
+};
+
+/**
+ * Tests whether a new price/quantity would cost less than the user's current subscription.
```

Comment:
:laughing: that definition of what a downgrade is - if it costs more it **must** be better.